src/aiida_quantumespresso/data/hubbard_structure.py

```diff
@@ -109,7 +109,16 @@ def append_hubbard_parameter(
         :param hubbard_type: hubbard type (U, V, J, ...), defaults to 'Ueff'
             (see :class:`~aiida_quantumespresso.common.hubbard.Hubbard` for full allowed values)
         """
-        pymat = self.get_pymatgen_structure()
+        if self.pbc != (True, True, True):
+            #create a temporary StructureData to obtain periodic sites.
+            ase = self.get_ase()
+            ase.set_pbc((True, True, True))
+            temp_structure = StructureData(ase=ase)
```

Comment:
I think it should be able to create a temp structure directly with StructureData (https://github.com/aiidateam/aiida-core/blob/06ea130df8854f621e25853af6ac723c37397ed0/src/aiida/orm/nodes/data/structure.py#L1592) instead of through ase structure. 

What will happened if the `pbc` is not 3D? It will raise an error or it will give the wrong results afterwards?

Reply:
if is different that 3D , it will try to use pymatgen, and then it will fail

Reply:
@unkcpz i think i cant use set_pbc since , is an StructureData that is saved

Reply:
you mean like this `temp_structure = StructureData(ase=ase, pbc=[True, True, True])` ?

Reply:
No I mean using as less structure type as possible. 
With your implementation, the structure type used are `HubbardStructureData` -> `asc structure` ->  `StructureData` -> pymatgen structure. However, the purpose here is to get the `sites` which are information that should include in the StructureData. 

I think we should try our best to not use ase/pymatgen, at least should try to not add more dependent on those. I want to see @mbercx's opinion on it.

Reply:
Thanks for the ping @unkcpz. 

I see your point that going through ASE to get a `StructureData` again to then get a `pymatgen` structure seems very roundabout, but I understand why @AndresOrtegaGuerrero did it. 

The issue is that `get_pymatgen()` would now work since it would return a pymatgen `Molecule`, and `get_pymatgen_structure()` fails since that only allows periodic boundary conditions... Ideally this _would_ actually be possible. I don't immediately understand why `StructureData` doesn't allow it, since a pymatgen `Structure` can also e.g. have `(True, False, False)` periodic boundary conditions. I'll open an issue on `aiida-core` to discuss this, but that'll take time.

Unfortunately, it's also not so easy to get an (unstored) copy of a `StructureData` directly. We could avoid using ASE in another way though. In the end the only thing the method needs is the sites, so the following should work for structures of any dimension:

```python
from pymatgen.core import PeriodicSite, Lattice

sites = [
    PeriodicSite(
        species=site.species,
        coords=site.coords,
        lattice=Lattice(self.cell),
        coords_are_cartesian=True
    ) for site in self.get_pymatgen().sites
]
```

And is probably cleaner than the if/else clause with the `StructureData` -> ASE -> `StructureData` -> `Structure` -> sites route for <3D structures.

As a final note, I know @unkcpz wants to avoid using ASE/pymatgen since these can be somewhat problematic dependencies (ASE because there are barely any releases, pymatgen since they often make breaking changes). But in the end I don't see a way to avoid using these tools, nor do I think it is desirable. That would involve reimplementing (and maintaining!) a lot of functionality ourselves.